src/Sarif/Comparers/ReportingDescriptorSortingComparer.cs

```diff
@@ -0,0 +1,170 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+/// <summary>
+/// All Comparer implementations should be replaced by auto-generated codes by JSchema as 
+/// part of EqualityComparer in a planned comprehensive solution.
+/// Tracking by issue: https://github.com/microsoft/jschema/issues/141
+/// </summary>
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    internal class ReportingDescriptorSortingComparer : IComparer<ReportingDescriptor>
+    {
+        internal static readonly ReportingDescriptorSortingComparer Instance = new ReportingDescriptorSortingComparer();
+
+        public int Compare(ReportingDescriptor left, ReportingDescriptor right)
```

Comment:
>Compare [](http://example.com/codeflow?start=19&length=7)

qurestion about design, do we need to compare all fields?
I see a few fields is not included like Relationships,MessageStrings

Reply:
This solution is a short term fix to help resolving issues some tools (E.g. PREfast) generate non-deterministic results for same targets run-over-run asap. 

It mainly focus on relevant types cause the non-determinisms we observed in the sample/test SARIF files. 

To compare all the fields we need to manually implement IComparer for all SARIF types. E.g. if compare Relationships, RelationshipComparer need to be implemented. 

In planned comprehensive solution, we will update JSchema to automatically generates IComparer code for all SARIF types instead of these hand-written code, at that time we will make sure all types are used for comparing. Created issue [https://github.com/microsoft/sarif-sdk/issues/2424](https://github.com/microsoft/sarif-sdk/issues/2424) for tracking.

Reply:
thanks for the detail explain.